## Gather

In [1]:
import pandas as pd
import numpy as np

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [3]:
patients.T

0    \
patient_id                                       1   
assigned_sex                                female   
given_name                                     Zoe   
surname                                    Wellish   
address                       576 Brown Bear Drive   
city                             Rancho California   
state                                   California   
zip_code                                     92390   
country                              United States   
contact       951-719-9170ZoeWellish@superrito.com   
birthdate                                7/10/1976   
weight                                       121.7   
height                                          66   
bmi                                           19.6   

                                                1    \
patient_id                                        2   
assigned_sex                                 female   
given_name                                   Pamela   
surname                                        Hill   
address                   2370 University Hill Road   
city                                      Armstrong   
state                                      Illinois   
zip_code                                      61812   
country                               United States   
contact       PamelaSHill@cuvox.de+1 (217) 569-3204   
birthdate                                  4/3/1967   
weight                                        118.8   
height                                           66   
bmi                                            19.2   

                                           2    \
patient_id                                   3   
assigned_sex                              male   
given_name                                 Jae   
surname                                 Debord   
address                  1493 Poling Farm Road   
city                                      York   
state                                 Nebraska   
zip_code                                 68467   
country                          United States   
contact       402-363-6804JaeMDebord@gustr.com   
birthdate                            2/19/1980   
weight                                   177.8   
height                                      71   
bmi                                       24.8   

                                                     3    \
patient_id                                             4   
assigned_sex                                        male   
given_name                                          Liêm   
surname                                             Phan   
address                              2335 Webster Street   
city                                          Woodbridge   
state                                                 NJ   
zip_code                                            7095   
country                                    United States   
contact       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
birthdate                                      7/26/1951   
weight                                             220.9   
height                                                70   
bmi                                                 31.7   

                                          4    \
patient_id                                  5   
assigned_sex                             male   
given_name                                Tim   
surname                               Neudorf   
address                  1428 Turkey Pen Lane   
city                                   Dothan   
state                                      AL   
zip_code                                36303   
country                         United States   
contact       334-515-7487TimNeudorf@cuvox.de   
birthdate                           2/18/1928   
weight                                  192.3   
height                                     27   
bmi                                      26.1   

                      

In [4]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0       veronika      jindrová  41u - 48u          -         7.63       7.20   
1         elliot    richardson          -  40u - 45u         7.56       7.09   
2       yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3           skye    gormanston  33u - 36u          -         7.97       7.62   
4         alissa        montez          -  33u - 29u         7.78       7.46   
5        jasmine         sykes          -  42u - 44u         7.56       7.18   
6         sophia        haugen  37u - 42u          -         7.65       7.27   
7          eddie        archer  31u - 38u          -         7.89       7.55   
8          saber        ménard          -  54u - 54u         8.08       7.70   
9           asia       woźniak  30u - 36u          -         7.76       7.37   
10        joseph           day  29u - 36u          -         7.70       7.19   
11     kristiina        hyypiä          -  36u - 38u         7.87       7.49   
12       roxanne     andreyeva  29u - 38u          -         9.54       9.14   
13        gregor          bole          -  47u - 45u         7.61       7.16   
14        simone  baumgaertner  27u - 37u          -         7.74       7.30   
15          enco        žibrik  55u - 68u          -         7.78       7.34   
16       camilla      zaitseva  28u - 37u          -         7.53       7.13   
17          gina          cain          -  36u - 36u         7.88       7.40   
18    addolorata      lombardi          -  49u - 46u         7.75       7.33   
19        khalid      johnsrud          -  54u - 54u         8.35       7.94   
20          mile        stanić          -  47u - 48u         7.66       7.24   
21         tekla       walczak  29u - 39u          -         7.61       7.29   
22   brancaleone         russo  53u - 60u          -         8.61       8.18   
23      chiemela     tobeolisa          -  43u - 47u         7.59       7.17   
24          isac          berg  31u - 41u          -         9.68       9.29   
25        benoît        bonami          -  44u - 43u         9.82       9.40   
26        suhaim         rahal          -  49u - 47u         7.94       7.50   
27        mizuki         iwata          -  45u - 46u         7.70       7.23   
28       clinton        miller  42u - 51u          -         7.79       7.40   
29        eugene       mironov  42u - 49u          -         7.81       7.48   
..           ...           ...        ...        ...          ...        ...   
250         chen           yao          -  56u - 57u         7.90       7.51   
251        aksel   vestergaard          -  42u - 38u         9.62       9.29   
252        ellen         luman          -  40u - 39u         9.27       8.77   
253       albino     schiavone  35u - 43u          -         7.56       7.15   
254         jose         combs          -  39u - 36u         7.89       7.42   
255       jia li          teng  48u - 54u          -         7.66       7.32   
256        ilija        horvat  42u - 50u          -         7.77       7.38   
257     mathilde      nørgaard          -  27u - 28u         8.50       8.10   
258       csilla      herczegh          -  43u - 46u         7.71       7.27   
259      aaliyah          rice          -  31u - 31u         7.64       7.33   
260        david      beauvais          -  26u - 23u         7.87       7.47   
261     caroline        shuler          -  50u - 54u         7.63       7.27   
262         alex      crawford  51u - 62u          -         7.69       7.30   
263      rebecca      jephcott  53u - 63u          -         7.96       7.57   
264   chukwumoge     ogochukwu          -  41u - 39u         7.95       7.56   
265       fearne      mcgregor          -  27u - 29u         7.83       7.48   
266       ursula         freud  42u - 54u          -         7.75       7.46   
267         leon        scholz          -  38u - 32u         7.72       7.29   
268   

In [5]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [8]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [9]:
# Find duplicated columns name accross our dataframes

In [10]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns

0           patient_id
1         assigned_sex
2           given_name
3              surname
4              address
5                 city
6                state
7             zip_code
8              country
9              contact
10           birthdate
11              weight
12              height
13                 bmi
14          given_name
15             surname
16             auralin
17             novodra
18         hba1c_start
19           hba1c_end
20        hba1c_change
21          given_name
22             surname
23    adverse_reaction
dtype: object

In [11]:
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [12]:
# Get patient df columns names 
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [13]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [14]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [15]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [16]:
patients.sample(5).T

202  \
patient_id                                     203   
assigned_sex                                female   
given_name                                  Jiřina   
surname                                   Šubrtová   
address                             4262 Heron Way   
city                                      Portland   
state                                           OR   
zip_code                                     97204   
country                              United States   
contact       JirinaSubrtova@rhyta.com503-820-7877   
birthdate                               12/10/1987   
weight                                       138.4   
height                                          61   
bmi                                           26.1   

                                                    460  \
patient_id                                          461   
assigned_sex                                       male   
given_name                                         Arne   
surname                                       Jørgensen   
address                               1033 Wayside Lane   
city                                            Oakland   
state                                                CA   
zip_code                                          94612   
country                                   United States   
contact       ArneJrgensen@fleckens.hu+1 (510) 238-1151   
birthdate                                     3/14/1998   
weight                                            163.5   
height                                               65   
bmi                                                27.2   

                                                     368  \
patient_id                                           369   
assigned_sex                                        male   
given_name                                         Corey   
surname                                         Nicholls   
address                       3427 Gerald L. Bates Drive   
city                                              Boston   
state                                                 MA   
zip_code                                            2110   
country                                    United States   
contact       CoreyNicholls@jourrapide.com1 617 830 7216   
birthdate                                      4/25/1989   
weight                                               165   
height                                                74   
bmi                                                 21.2   

                                                         65   \
patient_id                                                66   
assigned_sex                                            male   
given_name                                           Nicolas   
surname                                             Ferreira   
address                                  2310 Overlook Drive   
city                                               Rossville   
state                                                     IN   
zip_code                                               46065   
country                                        United States   
contact       NicolasSouzaFerreira@superrito.com765-379-3885   
birthdate                                         10/27/1976   
weight                                                 211.9   
height                                                    69   
bmi                                                     31.3   

                                                 140  
patient_id                                       141  
assigned_sex                                    male  
given_name                                      John  
surname                                  Teichelmann  
address                        631 Isaacs Creek Road  
city                                         Decatur  
state                                             IL  
zip_code                                       6

In [17]:
patients.surname.value_counts()

Doe            6
Taylor         3
Jakobsen       3
Johnson        2
Collins        2
Lương          2
Kadyrov        2
Schiavone      2
Correia        2
Batukayev      2
Nilsen         2
Silva          2
Aranda         2
Dratchev       2
Grímsdóttir    2
Parker         2
Gersten        2
Liễu           2
Lâm            2
Tucker         2
Cabrera        2
Cindrić        2
Ogochukwu      2
Bùi            2
Souza          2
Woźniak        2
Lund           2
Tạ             2
Hueber         2
Berg           2
              ..
Zaitseva       1
Lorenzo        1
Azuma          1
Hansen         1
Heimisson      1
Edwards        1
Markus         1
Chandler       1
Pérez          1
Fiorentino     1
Hori           1
Chang          1
Stanić         1
Nyborg         1
Eskelinen      1
Udinesi        1
Abraham        1
Herman         1
House          1
Blix           1
Zhdanov        1
Onwughara      1
Afamefuna      1
Sandgreen      1
Teichelmann    1
Boyles         1
Costa          1
Mironov       

In [18]:
patients.address.value_counts()

123 Main Street             6
648 Old Dear Lane           2
2778 North Avenue           2
2476 Fulton Street          2
3788 Worthington Drive      1
3402 Kildeer Drive          1
3977 Jail Drive             1
4015 Juniper Drive          1
4386 Badger Pond Lane       1
2645 Moore Avenue           1
4458 Stark Hollow Road      1
3683 Gorby Lane             1
2931 Romano Street          1
284 Mudlick Road            1
2166 Tipple Road            1
3558 Longview Avenue        1
465 Southern Street         1
2775 Single Street          1
2418 Smith Street           1
3697 Drainer Avenue         1
312 Jim Rosa Lane           1
3682 Stiles Street          1
3113 Timber Ridge Road      1
4320 Rardin Drive           1
685 Duncan Avenue           1
1434 Oak Ridge Drive        1
1275 Goldie Lane            1
2403 Ripple Street          1
1207 Garfield Road          1
4271 Cherry Ridge Drive     1
                           ..
3499 Baker Avenue           1
1027 Tenmile Road           1
4720 Gordo

In [19]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [20]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [21]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [22]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [23]:
sum(treatments.auralin.isnull())

0

In [24]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [25]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

#### `treatments`: Missing records (280 instead of 350)

##### Define
*Your definition here. Note: the missing `treatments` records are stored in a file named `treatments_cut.csv`, which you can see in this Jupyter Notebook's dashboard (click the **jupyter** logo in the top lefthand corner of this Notebook). Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) for the function used in the solution.*            

Import the cut treatments into a new df, and concatenete it with the main treatments df.

##### Code

In [26]:
# Your cleaning code here

In [27]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [28]:
treatments_clean.shape

(280, 7)

In [29]:
treatments_cut = pd.read_csv("treatments_cut.csv")
treatments_cut.head()

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jožka  resanovič  22u - 30u          -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u          -         7.85       7.45   
2      alwin   svensson  36u - 39u          -         7.78       7.34   
3        thể      lương          -  61u - 64u         7.64       7.22   
4     amanda    ribeiro  36u - 44u          -         7.85       7.47   

   hba1c_change  
0          0.34  
1           NaN  
2           NaN  
3          0.92  
4          0.38

In [30]:
treatments_clean = pd.concat([treatments_clean, treatments_cut], ignore_index=True)

##### Test

In [31]:
# Your testing code here
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [32]:
treatments_clean.shape

(350, 7)

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

##### Define
*Your definition here.*

Recalculate the `hba1c_change` column: `hba1c_start` minus `hba1c_end`. 

##### Code

In [33]:
# Your cleaning code here
treatments_clean["hba1c_change"] = treatments_clean["hba1c_start"] - treatments_clean["hba1c_end"]

##### Test

In [34]:
# Your testing code here
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

### Tidiness

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
*Your definition here.    
Hint 1: use regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). Here is an amazing [regex tutorial](https://regexone.com/).       
Hint 2: [various phone number regex patterns](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number).      
Hint 3: [email address regex pattern](http://emailregex.com/), which you might need to modify to distinguish the email from the phone number.*

* Extract the *phone number* and *email* variables from the *contact* column using regular expressions and pandas' `str.extract` method. 
* Drop the *contact* column when done.

##### Code

In [35]:
# Your cleaning code here

In [36]:
patients_clean["phone_number"] = patients_clean["contact"].str \
                                 .extract("((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})", expand=True)

patients_clean.head(3).T

0  \
patient_id                                       1   
assigned_sex                                female   
given_name                                     Zoe   
surname                                    Wellish   
address                       576 Brown Bear Drive   
city                             Rancho California   
state                                   California   
zip_code                                     92390   
country                              United States   
contact       951-719-9170ZoeWellish@superrito.com   
birthdate                                7/10/1976   
weight                                       121.7   
height                                          66   
bmi                                           19.6   
phone_number                          951-719-9170   

                                                  1  \
patient_id                                        2   
assigned_sex                                 female   
given_name                                   Pamela   
surname                                        Hill   
address                   2370 University Hill Road   
city                                      Armstrong   
state                                      Illinois   
zip_code                                      61812   
country                               United States   
contact       PamelaSHill@cuvox.de+1 (217) 569-3204   
birthdate                                  4/3/1967   
weight                                        118.8   
height                                           66   
bmi                                            19.2   
phone_number                      +1 (217) 569-3204   

                                             2  
patient_id                                   3  
assigned_sex                              male  
given_name                                 Jae  
surname                                 Debord  
address                  1493 Poling Farm Road  
city                                      York  
state                                 Nebraska  
zip_code                                 68467  
country                          United States  
contact       402-363-6804JaeMDebord@gustr.com  
birthdate                            2/19/1980  
weight                                   177.8  
height                                      71  
bmi                                       24.8  
phone_number                      402-363-6804

In [37]:
# [a-zA-Z] to signify emails in this dataset all start and end with letters
patients_clean["email"] = patients_clean["contact"].str \
                         .extract("([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])", expand=True)

patients_clean.head(3).T

0  \
patient_id                                       1   
assigned_sex                                female   
given_name                                     Zoe   
surname                                    Wellish   
address                       576 Brown Bear Drive   
city                             Rancho California   
state                                   California   
zip_code                                     92390   
country                              United States   
contact       951-719-9170ZoeWellish@superrito.com   
birthdate                                7/10/1976   
weight                                       121.7   
height                                          66   
bmi                                           19.6   
phone_number                          951-719-9170   
email                     ZoeWellish@superrito.com   

                                                  1  \
patient_id                                        2   
assigned_sex                                 female   
given_name                                   Pamela   
surname                                        Hill   
address                   2370 University Hill Road   
city                                      Armstrong   
state                                      Illinois   
zip_code                                      61812   
country                               United States   
contact       PamelaSHill@cuvox.de+1 (217) 569-3204   
birthdate                                  4/3/1967   
weight                                        118.8   
height                                           66   
bmi                                            19.2   
phone_number                      +1 (217) 569-3204   
email                          PamelaSHill@cuvox.de   

                                             2  
patient_id                                   3  
assigned_sex                              male  
given_name                                 Jae  
surname                                 Debord  
address                  1493 Poling Farm Road  
city                                      York  
state                                 Nebraska  
zip_code                                 68467  
country                          United States  
contact       402-363-6804JaeMDebord@gustr.com  
birthdate                            2/19/1980  
weight                                   177.8  
height                                      71  
bmi                                       24.8  
phone_number                      402-363-6804  
email                     JaeMDebord@gustr.com

In [38]:
# Note: axis=1 denotes that we are referring to a column, not a row
patients_clean = patients_clean.drop("contact", axis=1)

patients_clean.head(3).T

0                          1  \
patient_id                           1                          2   
assigned_sex                    female                     female   
given_name                         Zoe                     Pamela   
surname                        Wellish                       Hill   
address           576 Brown Bear Drive  2370 University Hill Road   
city                 Rancho California                  Armstrong   
state                       California                   Illinois   
zip_code                         92390                      61812   
country                  United States              United States   
birthdate                    7/10/1976                   4/3/1967   
weight                           121.7                      118.8   
height                              66                         66   
bmi                               19.6                       19.2   
phone_number              951-719-9170          +1 (217) 569-3204   
email         ZoeWellish@superrito.com       PamelaSHill@cuvox.de   

                                  2  
patient_id                        3  
assigned_sex                   male  
given_name                      Jae  
surname                      Debord  
address       1493 Poling Farm Road  
city                           York  
state                      Nebraska  
zip_code                      68467  
country               United States  
birthdate                 2/19/1980  
weight                        177.8  
height                           71  
bmi                            24.8  
phone_number           402-363-6804  
email          JaeMDebord@gustr.com

##### Test

In [39]:
patients_clean.head(3).T

0                          1  \
patient_id                           1                          2   
assigned_sex                    female                     female   
given_name                         Zoe                     Pamela   
surname                        Wellish                       Hill   
address           576 Brown Bear Drive  2370 University Hill Road   
city                 Rancho California                  Armstrong   
state                       California                   Illinois   
zip_code                         92390                      61812   
country                  United States              United States   
birthdate                    7/10/1976                   4/3/1967   
weight                           121.7                      118.8   
height                              66                         66   
bmi                               19.6                       19.2   
phone_number              951-719-9170          +1 (217) 569-3204   
email         ZoeWellish@superrito.com       PamelaSHill@cuvox.de   

                                  2  
patient_id                        3  
assigned_sex                   male  
given_name                      Jae  
surname                      Debord  
address       1493 Poling Farm Road  
city                           York  
state                      Nebraska  
zip_code                      68467  
country               United States  
birthdate                 2/19/1980  
weight                        177.8  
height                           71  
bmi                            24.8  
phone_number           402-363-6804  
email          JaeMDebord@gustr.com

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

In [40]:
treatments_clean.head(3)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43

##### Define
*Your definition here. Hint: use pandas' [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html). Here is an excellent [`melt` tutorial](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

* Melt the auralin and novodra columns to a treatment and a dose column (dose will still contain both start and end dose at this point). 
* Split the dose column on ' - ' to obtain start_dose and end_dose columns, while removing the dose unit "u" beside the values. 
* Drop the intermediate dose column.

##### Code

In [41]:
treatments_clean.columns.tolist()

['given_name',
 'surname',
 'auralin',
 'novodra',
 'hba1c_start',
 'hba1c_end',
 'hba1c_change']

In [42]:
id_vars = treatments_clean.columns.tolist()
id_vars.remove("auralin")
id_vars.remove("novodra")
id_vars

['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change']

In [43]:
melted = pd.melt(treatments_clean, id_vars=id_vars, var_name="treatment", value_name="dose")
melted.head(3)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
1     elliot  richardson         7.56       7.09          0.47   auralin   
2   yukitaka    takenaka         7.68       7.25          0.43   auralin   

        dose  
0  41u - 48u  
1          -  
2          -

In [44]:
melted.query(" given_name == '{}'".format("elliot"))

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
1       elliot  richardson         7.56       7.09          0.47   auralin   
351     elliot  richardson         7.56       7.09          0.47   novodra   

          dose  
1            -  
351  40u - 45u

In [45]:
melted["treatment"].unique()

array(['auralin', 'novodra'], dtype=object)

In [46]:
treatments_clean = melted[ melted["dose"]!="-" ]
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

        dose  
0  41u - 48u  
3  33u - 36u  
6  37u - 42u  
7  31u - 38u  
9  30u - 36u

In [47]:
treatments_clean["dose_start"], treatments_clean["dose_end"] = treatments_clean["dose"].str.replace("u","") \
                                                                                       .str.split(" - ", 1).str
treatments_clean.head()

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

        dose dose_start dose_end  
0  41u - 48u         41       48  
3  33u - 36u         33       36  
6  37u - 42u         37       42  
7  31u - 38u         31       38  
9  30u - 36u         30       36

In [48]:
treatments_clean.drop("dose", axis=1,inplace=True)
treatments_clean.head(2)

/opt/conda/lib/python3.6/site-packages/pandas/core/frame.py:3697: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   

  dose_start dose_end  
0         41       48  
3         33       36

##### Test

In [49]:
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  dose_start dose_end  
0         41       48  
3         33       36  
6         37       42  
7         31       38  
9         30       36

#### Adverse reaction should be part of the `treatments` table

In [50]:
adverse_reactions_clean.head(3)

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for the function used in the solution.*

* Merge the adverse_reaction column in adverse_reactions df to the treatments df, joining on given name and surname.

##### Code

In [51]:
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean, on=["given_name", "surname"], how='left')

##### Test

In [52]:
treatments_clean.head(3)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
1       skye  gormanston         7.97       7.62          0.35   auralin   
2     sophia      haugen         7.65       7.27          0.38   auralin   

  dose_start dose_end adverse_reaction  
0         41       48              NaN  
1         33       36              NaN  
2         37       42              NaN

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

In [53]:
patients_clean.head(2).T

0                          1
patient_id                           1                          2
assigned_sex                    female                     female
given_name                         Zoe                     Pamela
surname                        Wellish                       Hill
address           576 Brown Bear Drive  2370 University Hill Road
city                 Rancho California                  Armstrong
state                       California                   Illinois
zip_code                         92390                      61812
country                  United States              United States
birthdate                    7/10/1976                   4/3/1967
weight                           121.7                      118.8
height                              66                         66
bmi                               19.6                       19.2
phone_number              951-719-9170          +1 (217) 569-3204
email         ZoeWellish@superrito.com       PamelaSHill@cuvox.de

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for one function used in the solution and [tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) for another function used in the solution.*

* Adverse reactions df is no longer needed so ignore that part 
* Isolate the patient ID and names in the `patients` df, then convert these names to lower case to join with `treatments` df. 
* Then drop the given name and surname columns in the treatments df (so these being lowercase isn't an issue anymore).

##### Code

In [54]:
names = patients_clean[ ["patient_id","given_name","surname"] ]
names.head(2)

patient_id given_name  surname
0           1        Zoe  Wellish
1           2     Pamela     Hill

In [55]:
names["given_name"] = names["given_name"].str.lower()
names["surname"] = names["surname"].str.lower()
names.head(2)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


patient_id given_name  surname
0           1        zoe  wellish
1           2     pamela     hill

In [56]:
treatments_clean = pd.merge(treatments_clean, names, on=["given_name", "surname"])
treatments_clean.head(2)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
1       skye  gormanston         7.97       7.62          0.35   auralin   

  dose_start dose_end adverse_reaction  patient_id  
0         41       48              NaN         225  
1         33       36              NaN         242

In [57]:
treatments_clean.drop(["given_name", "surname"], axis=1,inplace=True)

##### Test

In [58]:
treatments_clean.head(3)

hba1c_start  hba1c_end  hba1c_change treatment dose_start dose_end  \
0         7.63       7.20          0.43   auralin         41       48   
1         7.97       7.62          0.35   auralin         33       36   
2         7.65       7.27          0.38   auralin         37       42   

  adverse_reaction  patient_id  
0              NaN         225  
1              NaN         242  
2              NaN         345

### Quality

#### Zip code is a float not a string and Zip code has four digits sometimes

In [59]:
patients_clean.head(3).T

0                          1  \
patient_id                           1                          2   
assigned_sex                    female                     female   
given_name                         Zoe                     Pamela   
surname                        Wellish                       Hill   
address           576 Brown Bear Drive  2370 University Hill Road   
city                 Rancho California                  Armstrong   
state                       California                   Illinois   
zip_code                         92390                      61812   
country                  United States              United States   
birthdate                    7/10/1976                   4/3/1967   
weight                           121.7                      118.8   
height                              66                         66   
bmi                               19.6                       19.2   
phone_number              951-719-9170          +1 (217) 569-3204   
email         ZoeWellish@superrito.com       PamelaSHill@cuvox.de   

                                  2  
patient_id                        3  
assigned_sex                   male  
given_name                      Jae  
surname                      Debord  
address       1493 Poling Farm Road  
city                           York  
state                      Nebraska  
zip_code                      68467  
country               United States  
birthdate                 2/19/1980  
weight                        177.8  
height                           71  
bmi                            24.8  
phone_number           402-363-6804  
email          JaeMDebord@gustr.com

##### Define

* Convert the zip code column's data type from a float to a string using astype, 
* Remove the '.0' using string slicing, 
* And pad four digit zip codes with a leading 0.

##### Code

In [60]:
patients_clean["zip_code"][:5]

0    92390.0
1    61812.0
2    68467.0
3     7095.0
4    36303.0
Name: zip_code, dtype: float64

In [61]:
zip_code = patients_clean["zip_code"].astype(str).str[:-2]
zip_code[:5]

0    92390
1    61812
2    68467
3     7095
4    36303
Name: zip_code, dtype: object

In [62]:
# https://www.w3resource.com/pandas/series/series-str-pad.php

patients_clean["zip_code"] = zip_code.str.pad(5, fillchar='0')
patients_clean["zip_code"][:5]

0    92390
1    61812
2    68467
3    07095
4    36303
Name: zip_code, dtype: object

In [63]:
# Reconvert NaNs entries that were converted to '0000n' by code above
patients_clean["zip_code"] = patients_clean["zip_code"].replace("0000n",np.nan)
patients_clean["zip_code"][:5]

0    92390
1    61812
2    68467
3    07095
4    36303
Name: zip_code, dtype: object

##### Test

In [64]:
patients_clean.head(5).T

0                          1  \
patient_id                           1                          2   
assigned_sex                    female                     female   
given_name                         Zoe                     Pamela   
surname                        Wellish                       Hill   
address           576 Brown Bear Drive  2370 University Hill Road   
city                 Rancho California                  Armstrong   
state                       California                   Illinois   
zip_code                         92390                      61812   
country                  United States              United States   
birthdate                    7/10/1976                   4/3/1967   
weight                           121.7                      118.8   
height                              66                         66   
bmi                               19.6                       19.2   
phone_number              951-719-9170          +1 (217) 569-3204   
email         ZoeWellish@superrito.com       PamelaSHill@cuvox.de   

                                  2                          3  \
patient_id                        3                          4   
assigned_sex                   male                       male   
given_name                      Jae                       Liêm   
surname                      Debord                       Phan   
address       1493 Poling Farm Road        2335 Webster Street   
city                           York                 Woodbridge   
state                      Nebraska                         NJ   
zip_code                      68467                      07095   
country               United States              United States   
birthdate                 2/19/1980                  7/26/1951   
weight                        177.8                      220.9   
height                           71                         70   
bmi                            24.8                       31.7   
phone_number           402-363-6804          +1 (732) 636-8246   
email          JaeMDebord@gustr.com  PhanBaLiem@jourrapide.com   

                                 4  
patient_id                       5  
assigned_sex                  male  
given_name                     Tim  
surname                    Neudorf  
address       1428 Turkey Pen Lane  
city                        Dothan  
state                           AL  
zip_code                     36303  
country              United States  
birthdate                2/18/1928  
weight                       192.3  
height                          27  
bmi                           26.1  
phone_number          334-515-7487  
email          TimNeudorf@cuvox.de

#### Tim Neudorf height is 27 in instead of 72 in

In [65]:
patients_clean[ patients_clean["height"]==27 ]

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      27  26.1   

   phone_number                email  
4  334-515-7487  TimNeudorf@cuvox.de

##### Define
Replace height for rows in the `patients` table that have a height of 27 in (there is only one) with 72 in.

##### Code

In [66]:
patients_clean["height"] = patients_clean["height"].replace(27, 72)

##### Test

In [67]:
patients_clean[ patients_clean["height"]==27 ]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

In [68]:
patients_clean[ patients_clean["surname"]=="Neudorf" ]

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      72  26.1   

   phone_number                email  
4  334-515-7487  TimNeudorf@cuvox.de

#### Full state names sometimes, abbreviations other times

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) for method used in solution.*

Apply a function that converts full state name to state abbreviation for California, New York, Illinois, Florida, and Nebraska.

##### Code

In [69]:
patients_clean["state"].unique()

array(['California', 'Illinois', 'Nebraska', 'NJ', 'AL', 'Florida', 'NV',
       'CA', 'MO', 'New York', 'MI', 'TN', 'VA', 'OK', 'GA', 'MT', 'MA',
       'NY', 'NM', 'IL', 'LA', 'PA', 'CO', 'ME', 'WI', 'SD', 'MN', 'FL',
       'WY', 'OH', 'IA', 'NC', 'IN', 'CT', 'KY', 'DE', 'MD', 'AZ', 'TX',
       'NE', 'AK', 'ND', 'KS', 'MS', 'WA', 'SC', 'WV', 'RI', 'NH', 'OR',
       nan, 'VT', 'ID', 'DC', 'AR'], dtype=object)

In [70]:
# Mapping from full state name to abbreviation
state_abbrev = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE'}

# Function to apply
def abbreviate_state(patient:pd.DataFrame):
    type(patient)
    if patient["state"] in state_abbrev.keys():
        abbrev = state_abbrev[patient["state"]]
        return abbrev
    else:
        return patient["state"]

In [71]:
patients_clean["state"] = patients_clean.apply(abbreviate_state, axis=1)

##### Test

In [72]:
patients_clean["state"].unique()

array(['CA', 'IL', 'NE', 'NJ', 'AL', 'FL', 'NV', 'MO', 'NY', 'MI', 'TN',
       'VA', 'OK', 'GA', 'MT', 'MA', 'NM', 'LA', 'PA', 'CO', 'ME', 'WI',
       'SD', 'MN', 'WY', 'OH', 'IA', 'NC', 'IN', 'CT', 'KY', 'DE', 'MD',
       'AZ', 'TX', 'AK', 'ND', 'KS', 'MS', 'WA', 'SC', 'WV', 'RI', 'NH',
       'OR', nan, 'VT', 'ID', 'DC', 'AR'], dtype=object)

#### Dsvid Gustafsson

In [73]:
patients_clean[ patients_clean["surname"] == "Gustafsson" ]

patient_id assigned_sex given_name     surname             address  \
8           9         male      Dsvid  Gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO    64105  United States  3/6/1937   163.9      66  26.5   

   phone_number                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

##### Define
Replace given name for rows in the `patients` df that have a given name of 'Dsvid' with 'David'.

##### Code

In [74]:
patients_clean["given_name"] = patients_clean["given_name"].replace("Dsvid", "David")

##### Test

In [75]:
patients_clean[ patients_clean["surname"] == "Gustafsson" ]

patient_id assigned_sex given_name     surname             address  \
8           9         male      David  Gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO    64105  United States  3/6/1937   163.9      66  26.5   

   phone_number                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

In [76]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null object
country         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
phone_number    491 non-null object
email           491 non-null object
dtypes: float64(2), int64(2), object(11)
memory usage: 59.0+ KB


In [77]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
hba1c_start         349 non-null float64
hba1c_end           349 non-null float64
hba1c_change        349 non-null float64
treatment           349 non-null object
dose_start          349 non-null object
dose_end            349 non-null object
adverse_reaction    35 non-null object
patient_id          349 non-null int64
dtypes: float64(3), int64(1), object(4)
memory usage: 24.5+ KB


In [78]:
treatments_clean.head(2)

hba1c_start  hba1c_end  hba1c_change treatment dose_start dose_end  \
0         7.63       7.20          0.43   auralin         41       48   
1         7.97       7.62          0.35   auralin         33       36   

  adverse_reaction  patient_id  
0              NaN         225  
1              NaN         242

##### Define
*Your definition here. Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html) for one method used in solution, [documentation page](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) for one function used in the solution, and [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html) for another method used in the solution.*

* Convert assigned sex and state to categorical data types. Zip code data type was already addressed above. 
* Convert birthdate to datetime data type. 
* Strip the letter 'u' in start dose and end dose (already addressed above) and convert those columns to data type integer.

##### Code

In [79]:
# To category
patients_clean["assigned_sex"] = patients_clean["assigned_sex"].astype('category')
patients_clean["state"] = patients_clean["state"].astype('category')
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null category
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null category
zip_code        491 non-null object
country         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
phone_number    491 non-null object
email           491 non-null object
dtypes: category(2), float64(2), int64(2), object(9)
memory usage: 53.9+ KB


In [80]:
# To datetime
patients_clean["birthdate"] = pd.to_datetime(patients_clean["birthdate"])
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null category
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null category
zip_code        491 non-null object
country         491 non-null object
birthdate       503 non-null datetime64[ns]
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
phone_number    491 non-null object
email           491 non-null object
dtypes: category(2), datetime64[ns](1), float64(2), int64(2), object(8)
memory usage: 53.9+ KB


In [81]:
# Strip u and convert to integer
treatments_clean["dose_start"] = treatments_clean["dose_start"].str.strip('u').astype(int)
treatments_clean["dose_end"] = treatments_clean["dose_end"].str.strip('u').astype(int)
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
hba1c_start         349 non-null float64
hba1c_end           349 non-null float64
hba1c_change        349 non-null float64
treatment           349 non-null object
dose_start          349 non-null int64
dose_end            349 non-null int64
adverse_reaction    35 non-null object
patient_id          349 non-null int64
dtypes: float64(3), int64(3), object(2)
memory usage: 24.5+ KB


##### Test

In [82]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null category
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null category
zip_code        491 non-null object
country         491 non-null object
birthdate       503 non-null datetime64[ns]
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
phone_number    491 non-null object
email           491 non-null object
dtypes: category(2), datetime64[ns](1), float64(2), int64(2), object(8)
memory usage: 53.9+ KB


In [83]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
hba1c_start         349 non-null float64
hba1c_end           349 non-null float64
hba1c_change        349 non-null float64
treatment           349 non-null object
dose_start          349 non-null int64
dose_end            349 non-null int64
adverse_reaction    35 non-null object
patient_id          349 non-null int64
dtypes: float64(3), int64(3), object(2)
memory usage: 24.5+ KB


#### Multiple phone number formats

In [84]:
patients_clean.head(3).T

0                          1  \
patient_id                           1                          2   
assigned_sex                    female                     female   
given_name                         Zoe                     Pamela   
surname                        Wellish                       Hill   
address           576 Brown Bear Drive  2370 University Hill Road   
city                 Rancho California                  Armstrong   
state                               CA                         IL   
zip_code                         92390                      61812   
country                  United States              United States   
birthdate          1976-07-10 00:00:00        1967-04-03 00:00:00   
weight                           121.7                      118.8   
height                              66                         66   
bmi                               19.6                       19.2   
phone_number              951-719-9170          +1 (217) 569-3204   
email         ZoeWellish@superrito.com       PamelaSHill@cuvox.de   

                                  2  
patient_id                        3  
assigned_sex                   male  
given_name                      Jae  
surname                      Debord  
address       1493 Poling Farm Road  
city                           York  
state                            NE  
zip_code                      68467  
country               United States  
birthdate       1980-02-19 00:00:00  
weight                        177.8  
height                           71  
bmi                            24.8  
phone_number           402-363-6804  
email          JaeMDebord@gustr.com

##### Define
*Your definition here. Hint: helpful [Stack Overflow answer](https://stackoverflow.com/a/123681).*

* Strip all " ", "-", "(", ")", and "+" and store each number without any formatting. 
* Pad the phone number with a 1 if the length of the number is 10 digits (we want country code).

##### Code

In [85]:
patients_clean["phone_number"] = patients_clean["phone_number"].str.replace(r'\D+', '').str.pad(11, fillchar='1')

##### Test

In [86]:
patients_clean.head(3).T

0                          1  \
patient_id                           1                          2   
assigned_sex                    female                     female   
given_name                         Zoe                     Pamela   
surname                        Wellish                       Hill   
address           576 Brown Bear Drive  2370 University Hill Road   
city                 Rancho California                  Armstrong   
state                               CA                         IL   
zip_code                         92390                      61812   
country                  United States              United States   
birthdate          1976-07-10 00:00:00        1967-04-03 00:00:00   
weight                           121.7                      118.8   
height                              66                         66   
bmi                               19.6                       19.2   
phone_number               19517199170                12175693204   
email         ZoeWellish@superrito.com       PamelaSHill@cuvox.de   

                                  2  
patient_id                        3  
assigned_sex                   male  
given_name                      Jae  
surname                      Debord  
address       1493 Poling Farm Road  
city                           York  
state                            NE  
zip_code                      68467  
country               United States  
birthdate       1980-02-19 00:00:00  
weight                        177.8  
height                           71  
bmi                            24.8  
phone_number            14023636804  
email          JaeMDebord@gustr.com

#### Default John Doe data

In [87]:
patients_clean.query(" given_name=='{}' & surname=='{}' ".format("John","Doe"))

patient_id assigned_sex given_name surname          address      city  \
215         216         male       John     Doe  123 Main Street  New York   
229         230         male       John     Doe  123 Main Street  New York   
237         238         male       John     Doe  123 Main Street  New York   
244         245         male       John     Doe  123 Main Street  New York   
251         252         male       John     Doe  123 Main Street  New York   
277         278         male       John     Doe  123 Main Street  New York   

    state zip_code        country  birthdate  weight  height   bmi  \
215    NY    12345  United States 1975-01-01   180.0      72  24.4   
229    NY    12345  United States 1975-01-01   180.0      72  24.4   
237    NY    12345  United States 1975-01-01   180.0      72  24.4   
244    NY    12345  United States 1975-01-01   180.0      72  24.4   
251    NY    12345  United States 1975-01-01   180.0      72  24.4   
277    NY    12345  United States 1975-01-01   180.0      72  24.4   

    phone_number              email  
215  11234567890  johndoe@email.com  
229  11234567890  johndoe@email.com  
237  11234567890  johndoe@email.com  
244  11234567890  johndoe@email.com  
251  11234567890  johndoe@email.com  
277  11234567890  johndoe@email.com

##### Define
*Your definition here. Recall that it is assumed that the data that this John Doe data displaced is not recoverable.*

* Remove the non-recoverable John Doe records from the patients table.

##### Code

In [88]:
patients_clean = patients_clean.query(" given_name!='{}' & surname!='{}' ".format("John","Doe"))

##### Test

In [89]:
patients_clean.query(" given_name=='{}' & surname=='{}' ".format("John","Doe"))

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

#### Multiple records for Jakobsen, Gersten, Taylor

In [90]:
patients_clean[ (patients_clean.address.duplicated()) & patients_clean.address.notnull() ]

patient_id assigned_sex given_name   surname             address  \
29           30         male       Jake  Jakobsen   648 Old Dear Lane   
282         283       female      Sandy    Taylor  2476 Fulton Street   
502         503         male        Pat   Gersten   2778 North Avenue   

            city state zip_code        country  birthdate  weight  height  \
29   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
282     Rainelle    WV    25962  United States 1960-10-23   206.1      64   
502         Burr    NE    68324  United States 1954-05-03   138.2      71   

      bmi phone_number                      email  
29   24.4  18458587707  JakobCJakobsen@einrot.com  
282  35.4  13044382648   SandraCTaylor@dayrep.com  
502  19.3  14028484923   PatrickGersten@rhyta.com

##### Define

Remove the Jake Jakobsen, Pat Gersten, and Sandy Taylor rows from the `patients` df. These are the nicknames, which happen to also not be in the `treatments` df (removing the wrong name would create a consistency issue between the `patients` and `treatments` df).  These are all the second occurrence of the duplicate. These are also the only occurences of non-null duplicate addresses.

##### Code

In [91]:
# tilde means not: http://pandas.pydata.org/pandas-docs/stable/indexing.html#boolean-indexing
patients_clean = patients_clean[~( (patients_clean.address.duplicated()) & patients_clean.address.notnull() )]

##### Test

In [92]:
patients_clean[ (patients_clean.address.duplicated()) & patients_clean.address.notnull() ]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

#### kgs instead of lbs for Zaitseva weight

In [93]:
patients_clean[ patients_clean["surname"] == "Zaitseva" ]

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state zip_code        country  birthdate  weight  height   bmi  \
210  Wooster    OH    44691  United States 1938-11-26    48.8      63  19.1   

    phone_number                          email  
210  13302022145  CamillaZaitseva@superrito.com

##### Define
Use [advanced indexing](https://stackoverflow.com/a/44913631) to isolate the row where the surname is Zaitseva and convert the entry in its weight field from kg to lbs.

##### Code

In [94]:
weight_kg = patients_clean[ patients_clean["surname"] == "Zaitseva" ].loc[210,"weight"]
weight_kg

48.799999999999997

In [95]:
patients_clean.loc[ patients_clean["surname"] == "Zaitseva" , "weight"]

210    48.8
Name: weight, dtype: float64

In [96]:
patients_clean.loc[ patients_clean["surname"] == "Zaitseva" , "weight"] = weight_kg * 2.20462

##### Test

In [97]:
patients_clean[ patients_clean["surname"] == "Zaitseva" ]

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state zip_code        country  birthdate      weight  height  \
210  Wooster    OH    44691  United States 1938-11-26  107.585456      63   

      bmi phone_number                          email  
210  19.1  13302022145  CamillaZaitseva@superrito.com